datadotworld/files.py

```diff
@@ -99,22 +104,73 @@ def write(self, value):
                 self._queue.put(value.encode('utf-8'))
             else:
                 raise TypeError(
-                    "write() argument must be str, not {}"
-                        .format(type(value)))
+                    "write() argument must be str, not {}".format(
+                        type(value)))
         elif 'wb' == self._mode:
-            if (isinstance(value, (bytes, bytearray))):
+            if isinstance(value, (bytes, bytearray)):
                 self._queue.put(value)
             else:
                 raise TypeError(
-                    "a bytes-like object is required, not {}"
-                        .format(type(value)))
+                    "a bytes-like object is required, not {}".format(
+                        type(value)))
         else:
             raise NotImplementedError(
-                "modes other than 'w' and 'wb' not supported")
+                "write only available for files opened in write mode")
+
+    def read(self):
+        """
+        read the contents of the file that's been opened in read mode
+        """
+        if self._mode.find('r') >= 0:
+            return self._read_response.text.encode('utf-8')
+        else:
+            raise NotImplementedError(
+                "read only available for files opened in read mode")
+
+    def __iter__(self):
+        """
+        the file object must be able to act as an iterable of lines to be
+        used similarly to how a local file would be used e.g. in CSV
+        reading
+        """
+        return self._read_response.iter_lines()
 
     def open(self):
         """
-        start the thread executing the HTTP request
+        in write mode, start the thread executing the HTTP request.  in read
+        mode, execute the GET request and hold on to the response.
+        """
+        if self._mode.find('w') >= 0:
+            self._open_for_write()
+        else:
+            self._open_for_read()
+
+    def _open_for_read(self):
+        """
+        open the file in read mode
+        """
+        ownerid, datasetid = parse_dataset_key(self._dataset_key)
+        response = requests.get(
+            '{}/file_download/{}/{}/{}'.format(
+                self._query_host, ownerid, datasetid, self._file_name),
+            headers={
+                'Authorization': 'Bearer {}'.format(
+                    self._config.auth_token)
+            }, stream=True)
```

Comment:
@rflprr  - turns out adding streaming was as simple as this two-line change:

https://github.com/datadotworld/data.world-py/pull/62/commits/e6911416d3be87cca7157635d549e7caaa397209